### ISSUE TITLE
Inconsistencies comparing `self` and `other`

### ISSUE THREAD

**Issue by backYard321** [YoYoGames/GameMaker-Bugs#4788]
### Description

In a few cases, checking `self == other;` and `variable = self; variable == other;` will give two different results when used in the same context.

[self.and.other.zip](https://github.com/YoYoGames/GameMaker-Bugs/files/14391298/self.and.other.zip)

### Expected Change

`self == other` should be true if checking a variable storing the same self would also be true.

### Steps To Reproduce

1. Start GameMaker & create new project
2. Create an object, and write the following in the create event:
```gml
show_message(self == other);
var _variable = self;
show_message(_variable == other);
```
3. Place the object in the room
4. Run the game and see the issue - the first statement will show `0`, while the second will show `1`, even though they should be functionally identical (if I understand correctly).

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2023.11.1 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [X] I have included a small sample project

**Comment by nullbunny** [comment]
Since self is a struct now, you might have to check self.id - I could be wrong tho.

**Comment by backYard321** [comment]
> Since self is a struct now, you might have to check self.id - I could be wrong tho.

While I appreciate the tip, I already have a workaround for this - I still think the fundamental problem is a bug that should be addressed.

**Comment by attic-stuff** [comment]
there is a bug here, self should not equal other in this context because **self** is the self struct of Object1 and **other** should have no context here. functionally both of those are empty structs.

that being said, you are checking if two structs are the same struct (which means, does this have the same pointer as the other?) and ultimately i cannot think of a single situation where self and other would be contextually the same struct!

**Comment by rwkay** [comment]
Fixed in 2024.4 - Constant Optimiser was treating `self`, `other and `global` as constant numbers and not as references, it has been instructed to stop doing that in this modern age

**Comment by backYard321** [comment]
> Fixed in 2024.4 - Constant Optimiser was treating `self`, `other and `global` as constant numbers and not as references, it has been instructed to stop doing that in this modern age

![tumblr_19a2e4144afa8b523efc36afdce42f90_d97c86ba_500](https://github.com/YoYoGames/GameMaker-Bugs/assets/40671879/7f38185b-a847-4885-aa7a-be692fe65810)

**Comment by mgeddesGM** [comment]
verified as of Beta IDE v2024.400.0.532  Runtime v2024.400.0.551
